### PR TITLE
"Simple" implementation of fork choice rule

### PR DESCRIPTION
**Pull request by mratsim**
This is a simple implementation of LMD Ghost without justified and finalized head considerations. Next step is to call it at https://github.com/status-im/nim-beacon-chain/blob/96bf7e377ac6b9a422f7459309d9f865ed8e06e7/beacon_chain/beacon_node.nim#L293-L317.

References:
  - Specs: https://github.com/ethereum/eth2.0-specs/blob/197b88df9272317db5ef8adf730dc490a1e7180e/specs/core/0_beacon-chain.md#beacon-chain-fork-choice-rule
  - Mini-specs: https://ethresear.ch/t/beacon-chain-casper-mini-spec/2760
  - Research implementation: https://github.com/ethereum/research/blob/3aa84a76dd7cf4a003876d2766ab4aa1726f2622/clock_disparity/lmd_node.py
  - Nim implementation of IMD Ghost (not LMD): https://github.com/status-im/nim-beacon-chain/blob/96bf7e377ac6b9a422f7459309d9f865ed8e06e7/research/fork_choice_rule/ethresearch/fork_choice_rule.nim

Furthermore while the spec tracks vote counts per validator we track counts per shard/commitee so `get_latest_attestation(store, validator)` and `get_latest_attestation_target(store, validator)` are not implemented.

Things that are missing and I added in a raw/stub form:

- participation_bitfield: a way to get vote count as in #19 and https://github.com/ethereum/beacon_chain/blob/fd86633467f0cedaba838201d98a2ddcfdedc7db/beacon_chain/utils/bitfield.py#L21-L26
- A mapping between Eth2Digest and BeaconBlock, added in beacon_chain_db: https://github.com/status-im/nim-beacon-chain/blob/96bf7e377ac6b9a422f7459309d9f865ed8e06e7/beacon_chain/beacon_chain_db.nim#L9
- A way to track children of a block, similar to https://github.com/ethereum/research/blob/3aa84a76dd7cf4a003876d2766ab4aa1726f2622/clock_disparity/lmd_node.py#L84, the spec mentions `get_children` but doesn't propose an implementation or even a data structure. Currently the children are added and reset at the beacon node level similar to ethresear.ch implementation (not implemented yet): https://github.com/status-im/nim-beacon-chain/blob/96bf7e377ac6b9a422f7459309d9f865ed8e06e7/beacon_chain/beacon_node.nim#L22 and https://github.com/status-im/nim-beacon-chain/blob/96bf7e377ac6b9a422f7459309d9f865ed8e06e7/research/fork_choice_rule/ethresearch/fork_choice_types.nim#L129 but it might be better to track them in the AttestationPool

Future: The current head recalculation is O(N^2) and can be log(N) by using [Fenwick Trees](https://en.wikipedia.org/wiki/Fenwick_tree). I have a [Nim implementation for cumulative probabilities sampling](https://github.com/numforge/laser/blob/990e59fffe50779cdef33aa0b8f22da19e1eb328/benchmarks/random_sampling/fenwicktree.nim) that can be adapted to cumulative sums.

### CODE REVIEW

**Comment by mratsim** [comment]
This is ready for a first round of review.

I will add smoke tests afterward
And integrate it in the beacon node in a third step.

File: beacon_chain/beacon_node.nim [hunk 1]
```diff
@@ -3,7 +3,8 @@ import
   asyncdispatch2, chronicles, confutils, eth_p2p, eth_keys,
   spec/[datatypes, digest, crypto, beaconstate, helpers], conf, time,
   state_transition, fork_choice, ssz, beacon_chain_db, validator_pool, extras,
-  mainchain_monitor, sync_protocol, gossipsub_protocol, trusted_state_snapshots
+  mainchain_monitor, sync_protocol, gossipsub_protocol, trusted_state_snapshots,
+  tables
```
Comment (zah) [inline beacon_chain/beacon_node.nim]: BTW, I usually try to keep the import blocks sorted in the following order:

```nim
import
  std_modules,
  nimble_packages,
  local_project_files_in_other_directories, local_project_files_in_the_current_dir
```

File: beacon_chain/fork_choice.nim [hunk 2]
```diff
@@ -108,13 +109,121 @@ func getAttestationCandidate*(attestation: Attestation): AttestationCandidate =
...
+  else:
+    return store.get_ancestor(store.get_parent(blck), slot)
+  # TODO: what if the slot was never observed/verified?
+
+func getVoteCount(participation_bitfield: openarray[byte]): int =
+  ## Get the number of votes
+  # TODO: A bitfield type that tracks that information
+  # https://github.com/status-im/nim-beacon-chain/issues/19
+
+  for validatorIdx in 0 ..< participation_bitfield.len * 8:
```
Comment (zah) [inline beacon_chain/fork_choice.nim]: Most modern CPUs have a built-in instruction for getting the number of raised bits in a word:
https://en.wikipedia.org/wiki/Hamming_weight

In C, these are exposed as intrinsic functions:
https://stackoverflow.com/questions/3849337/msvc-equivalent-to-builtin-popcount

I guess It's best if we add helpers for using the intrinsics in nim-ranges/bitranges.

Comment (tersec) [reply beacon_chain/fork_choice.nim]: Sure, though this is just one more special case of lots of kludgy workarounds in this code for not using `nim-ranges` or something similar. Not accessing bit-counting instructions is merely one of the costs of this.

**Comment by zah** [comment]
The CI failure is related to using `sizeof(T)` at compile time:

```
spec/digest.nim(56, 17) Error: cannot run in the VM: sizeof(x)
```

The support for this landed recently, it's not present in 0.19.0.

File: research/fork_choice_rule/lmd_ghost.nim [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+  # `seq[AttestationSignedData]` is a queue ordered by (slot, observation time)
+
+  let attestations = store.verified_attestations[validatorIdx]
+  result = attestations[^1]                        # Pick the last attestation
+  for idx in countdown(attestations[].len - 2, 0): # From the second to last attestation to 0, check if they have the same slot.
+    if attestations[idx].slot == result.slot:      # if yes it was observed earlier
+      result = attestations[idx]
+    else:                                          # otherwise we are at the first observed attestation with the highest slot
+      return
+
```
[the remark this answers is not in the document]
Comment (mratsim) [reply research/fork_choice_rule/lmd_ghost.nim]: As mentionned in the Gitter, block processing spec used to be the same before being more fleshed out: https://github.com/ethereum/eth2.0-specs/blob/0001b7b9de2bf87ff267547acdb99788cf9b463c/specs/casper_sharding_v2.1.md#per-block-processing

So I guess this is a placeholder until we have runnable code.

**Comment by zah** [comment]
On the last call, I mentioned that I'm planning to add few more TODO items in the fork_choice module, but it turned out everything is already there :)

**Pull request merged**